diff --git a/mlx/backend/cuda/worker.cpp b/mlx/backend/cuda/worker.cpp
--- a/mlx/backend/cuda/worker.cpp
+++ b/mlx/backend/cuda/worker.cpp
@@ -57,9 +57,9 @@ void Worker::commit(cudaStream_t stream) {
   CHECK_CUDA_ERROR(cudaLaunchHostFunc(signal_stream_, signal, this));
 }
 
 void Worker::thread_fn() {
-  while (!stop_) {
-    uint64_t current_batch = 0;
+  uint64_t current_batch = 0;
+  while (true) {
     Tasks tasks;
     {
       std::unique_lock<std::mutex> lk(mtx_);
@@ -66,6 +66,9 @@ void Worker::thread_fn() {
       cond_.wait(lk, [this, &current_batch] {
         return this->signaled_batch_ > current_batch || this->stop_;
       });
+      if (stop_) {
+        return;
+      }
       current_batch = signaled_batch_;
       auto end = worker_tasks_.upper_bound(current_batch);
       for (auto it = worker_tasks_.begin(); it != end; ++it) {
